extensions/sql-database-projects/src/dialogs/addDatabaseReferenceDialog.ts

```diff
@@ -0,0 +1,402 @@
+							variableSection
+						]
+					}
+				], {
+					horizontal: false
+				})
+				.withLayout({
+					width: '100%'
+				});
+
+			let formModel = this.formBuilder.component();
+			await view.initializeModel(formModel);
+
+			this.initDialogComplete?.resolve();
+		});
+	}
+
+	public async addReferenceClick(): Promise<void> {
+		let referenceSettings: ISystemDatabaseReferenceSettings | IDacpacReferenceSettings;
+
+		if (this.currentReferenceType === ReferenceType.systemDb) {
+			referenceSettings = {
+				databaseName: <string>this.databaseNameTextbox?.value,
+				systemDb: <string>this.systemDatabaseDropdown?.value === constants.master ? SystemDatabase.master : SystemDatabase.msdb
+			};
+		} else { // this.currentReferenceType === ReferenceType.dacpac
+			referenceSettings = {
+				databaseName: <string>this.databaseNameTextbox?.value,
+				databaseLocation: <DatabaseReferenceLocation>this.referenceLocationMap.get(<string>this.locationDropdown?.value),
+				dacpacFileLocation: vscode.Uri.file(<string>this.dacpacTextbox?.value),
+				databaseVariable: <string>this.databaseVariableTextbox?.value
+			};
+			// TODO: add project reference support
+		}
+
+		await this.addReference!(this.project, referenceSettings);
+
+		this.dispose();
+	}
+
+	private createRadioButtons(): azdata.FormComponent {
+		// TODO: add project reference button
+		const systemDatabaseRadioButton = this.view!.modelBuilder.radioButton()
+			.withProperties({
+				name: 'referenceType',
+				label: constants.systemDatabaseRadioButtonTitle
+			}).component();
+
+		systemDatabaseRadioButton.checked = true;
+		systemDatabaseRadioButton.onDidClick(() => {
+			this.systemDbRadioButtonClick();
+		});
+
+		const dacpacRadioButton = this.view!.modelBuilder.radioButton()
+			.withProperties({
+				name: 'referenceType',
+				label: constants.dacpacText
+			}).component();
+
+		dacpacRadioButton.onDidClick(() => {
+			this.dacpacRadioButtonClick();
+		});
+
+		this.currentReferenceType = ReferenceType.systemDb;
+		let flexRadioButtonsModel: azdata.FlexContainer = this.view!.modelBuilder.flexContainer()
+			.withLayout({ flexFlow: 'column' })
+			.withItems([systemDatabaseRadioButton, dacpacRadioButton])
+			.withProperties({ ariaRole: 'radiogroup' })
+			.component();
+
+		return {
+			component: flexRadioButtonsModel,
+			title: constants.referenceRadioButtonsGroupTitle
+		};
+	}
+
+	public systemDbRadioButtonClick(): void {
+		this.formBuilder!.removeFormItem(<azdata.FormComponent>this.dacpacFormComponent);
+		this.formBuilder!.insertFormItem(<azdata.FormComponent>this.systemDatabaseFormComponent, 2);
+
+		// update dropdown values because only different database, same server is a valid location for system db references
+		this.locationDropdown!.values = constants.systemDbLocationDropdownValues;
+
+		this.currentReferenceType = ReferenceType.systemDb;
+		this.updateEnabledInputBoxes(true);
+		this.tryEnableAddReferenceButton();
+	}
+
+	public dacpacRadioButtonClick(): void {
```

Comment:
Are these set to be public just for testing purposes?

Reply:
yup just to make testing easier